File: solutions/Java/SummaryRangesSimulation.java
import java.util.ArrayList;
import java.util.List;

public class SummaryRangesSimulation {
    public List<String> summaryRanges(int[] nums) {
        /*
        Simulation approach: iterate through and build ranges as we go.
        Time: O(n), Space: O(1) excluding output
        */
        List<String> result = new ArrayList<>();
        if (nums.length == 0) return result;

        long start = nums[0];

        for (int i = 1; i < nums.length; i++) {
            if (nums[i] != nums[i - 1] + 1) {
                if (start == nums[i - 1]) {
                    result.add(String.valueOf(start));
                } else {
                    result.add(start + "->" + nums[i - 1]);
                }
                start = nums[i];
            }
        }

        if (start == nums[nums.length - 1]) {
            result.add(String.valueOf(start));
        } else {
            result.add(start + "->" + nums[nums.length - 1]);
        }

        return result;
    }
}

File: solutions/Python/summary_ranges_simulation.py
from typing import List


def summary_ranges(nums: List[int]) -> List[str]:
    """
    Simulation approach: iterate through and build ranges as we go.
    Time: O(n), Space: O(1) excluding output
    """
    if not nums:
        return []

    result = []
    start = nums[0]

    for i in range(1, len(nums)):
        # If current number is not consecutive, finalize the range
        if nums[i] != nums[i - 1] + 1:
            if start == nums[i - 1]:
                result.append(str(start))
            else:
                result.append(f"{start}->{nums[i - 1]}")
            start = nums[i]

    # Add the last range
    if start == nums[-1]:
        result.append(str(start))
    else:
        result.append(f"{start}->{nums[-1]}")

    return result


print(summary_ranges([0, 1, 2, 4, 5, 7]))  # ["0->2","4->5","7"]
print(summary_ranges([0, 2, 3, 4, 6, 8, 9]))  # ["0","2->4","6","8->9"]
